### PR TITLE
Nox system test refactor

### PR DESCRIPTION
**Pull request by jonparrott**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,9 +17,9 @@ matrix:
   - python: 3.5
     env: TOXENV=cover
   - python: 3.5
-    env: TOXENV=py35-system SYSTEM_TEST=1 SKIP_APP_ENGINE_SYSTEM_TEST=1
+    env: TOXENV=py35-system SYSTEM_TEST=1 SKIP_APP_ENGINE_SYSTEM_TEST=1 CLOUD_SDK_ROOT=${HOME}/.cache/cloud-sdk
   - python: 2.7
-    env: TOXENV=py27-system SYSTEM_TEST=1 SKIP_APP_ENGINE_SYSTEM_TEST=1
+    env: TOXENV=py27-system SYSTEM_TEST=1 SKIP_APP_ENGINE_SYSTEM_TEST=1 CLOUD_SDK_ROOT=${HOME}/.cache/cloud-sdk
```
Comment (dhermes) [inline .travis.yml]: Starting to become unwieldy. Can they go on multiple lines?

**Comment by jonparrott** [comment]
Not sure why travis is failing, I may have somehow broken the secrets. Will investigate tomorrow.

**Comment by jonparrott** [comment]
>  Not sure why travis is failing, I may have somehow broken the secrets. Will investigate tomorrow.

nvm, I think I did something dumb so it's maybe fixed now. :)

File: system_tests/nox.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+HERE = os.path.dirname(__file__)
+DATA_DIR = os.path.join(HERE, 'data')
+SERVICE_ACCOUNT_FILE = os.path.join(DATA_DIR, 'service_account.json')
+AUTHORIZED_USER_FILE = os.path.join(DATA_DIR, 'authorized_user.json')
+CLOUD_SDK_CONFIG_ENV = 'CLOUDSDK_CONFIG'
+
+# If set, this is where the environment setup will store the Cloud SDK.
+# If unset, it will download the SDK to a temporary directory.
+CLOUD_SDK_ROOT = os.environ.get('CLOUD_SDK_ROOT')
+if CLOUD_SDK_ROOT:
```
Comment (dhermes) [inline system_tests/nox.py]: `CLOUD_SDK_ROOT is not None`, there goes that @jonparrott singleton-allergy again

File: system_tests/nox.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    CLOUD_SDK_ROOT = py.path.local(CLOUD_SDK_ROOT)
```
Comment (dhermes) [inline system_tests/nox.py]: Refresh me on what `py.path.local` does

File: system_tests/nox.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    CLOUD_SDK_ROOT.ensure(dir=True)
+if not CLOUD_SDK_ROOT:
```
Comment (dhermes) [inline system_tests/nox.py]: Why not an `else` branch? Does` ensure(dir=True)` guarantee creation?

File: system_tests/nox.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+
+# Cloud SDK helpers
+
+
+def setup_cloud_sdk():
+    """Downloads and installs the Google Cloud SDK."""
+
+    # If the sdk already exists, we don't need to do anything else.
+    if CLOUD_SDK_ROOT.join('google-cloud-sdk').exists():
```
Comment (dhermes) [inline system_tests/nox.py]: Make `'google-cloud-sdk'` a constant

File: system_tests/nox.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        return
```
Comment (dhermes) [inline system_tests/nox.py]: Make a note that you intentionally don't check that it's updated, only that it exists (don't you want to check that it is a directory?)

File: system_tests/nox.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+    tar_file = 'google-cloud-sdk.tar.gz'
+    tar_path = CLOUD_SDK_ROOT.join(tar_file)
+
+    # Download the release.
+    prerun(
+        'wget', 'https://dl.google.com/dl/cloudsdk/release/{}'.format(
```
Comment (dhermes) [inline system_tests/nox.py]: Make the `dl.google.com` template and the `tar.gz` filename constants?

File: system_tests/nox.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+            tar_file),
+        '-O', str(tar_path))
+
+    # Extract the release.
+    prerun('tar', 'xzf', str(tar_path), '-C', str(CLOUD_SDK_ROOT))
```
Comment (dhermes) [inline system_tests/nox.py]: What's the `pre` in `prerun`? It's just more like `pre_and_run`

File: system_tests/nox.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    tar_path.remove()
+
+    # Run the install script.
+    prerun(
+        str(CLOUD_SDK_ROOT.join('google-cloud-sdk', 'install.sh')),
```
Comment (dhermes) [inline system_tests/nox.py]: This is scary

File: system_tests/nox.py [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+    # unfortunately gcloud doesn't provide a clean way to tell it to use
+    # a particular set of credentials. However, this does verify that gcloud
+    # also considers the credentials valid by calling application-default
+    # print-access-token
+    dest = CLOUD_SDK_ROOT.join('application_default_credentials.json')
+    if dest.exists():
+        dest.remove()
+    py.path.local(application_default_credentials).copy(dest)
+
+    gcloud('auth', 'application-default', 'print-access-token')
```
Comment (dhermes) [inline system_tests/nox.py]: What does this do?

File: system_tests/nox.py [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+def session_oauth2_credentials(session):
+    session.virtualenv = False
+    session.run('pytest', 'test_oauth2_credentials.py')
+
+
+def session_default_explicit_service_account(session):
+    session.virtualenv = False
+    session.env['GOOGLE_APPLICATION_CREDENTIALS'] = SERVICE_ACCOUNT_FILE
+    session.env['EXPECT_PROJECT_ID'] = '1'
```
Comment (dhermes) [inline system_tests/nox.py]: Make these env. vars. constants

File: system_tests/nox.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    session.run('pytest', 'test_default.py')
+
+
+def session_default_cloud_sdk_authorized_user(session):
+    session.virtualenv = False
+    setup_cloud_sdk()
+    configure_cloud_sdk(AUTHORIZED_USER_FILE)
+
+    session.env[CLOUD_SDK_CONFIG_ENV] = str(CLOUD_SDK_ROOT)
+    session.run('pytest', '--pdb', 'test_default.py')
```
Comment (dhermes) [inline system_tests/nox.py]: Isn't `--pdb` gonna block on failures?

File: .travis.yml (hunk 1 above)
Comment (jonparrott) [inline .travis.yml]: Not really, but I'm gonna try to move them back to global.

File: system_tests/nox.py (hunk 3 above)
Comment (jonparrott) [inline system_tests/nox.py]: Just wraps a path with helpful methods like 'exists', 'ensure', etc.

File: system_tests/nox.py (hunk 4 above)
Comment (jonparrott) [inline system_tests/nox.py]: else done. `ensure` does promise creation.

File: system_tests/nox.py (hunk 5 above)
Comment (jonparrott) [inline system_tests/nox.py]: Done.

File: system_tests/nox.py (hunk 6 above)
Comment (jonparrott) [inline system_tests/nox.py]: Done.

File: system_tests/nox.py (hunk 7 above)
Comment (jonparrott) [inline system_tests/nox.py]: Done.

File: system_tests/nox.py (hunk 8 above)
Comment (jonparrott) [inline system_tests/nox.py]: I actually just removed this. I found a way to get nox to run everything in the session instead of trying to do it before the session runs.

File: system_tests/nox.py (hunk 9 above)
Comment (jonparrott) [inline system_tests/nox.py]: I know.

File: system_tests/nox.py (hunk 11 above)
Comment (jonparrott) [inline system_tests/nox.py]: Done.

File: system_tests/nox.py (hunk 12 above)
Comment (jonparrott) [inline system_tests/nox.py]: Yes, leftovers, removed.

**Comment by jonparrott** [comment]
@dhermes I did a bit of refactoring to remove the hacky usage of `prerun`, PTAL.

File: .travis.yml [hunk 13]
```diff
@@ -38,4 +38,5 @@ deploy:
...
+    - secure: s6GdhJklftl8w/9WoETwLtvtKL4ledPA/TuBuqCXQxSuYWaPuTdRVcvoejGkHJpp7i/7v2T/0etYl+5koyskKm5+QZZweaaL7MAyjPGp+hmIaIlWQRz6w481NOf3i9uSmoQycssT0mNmwScNIqo+igbA2y14mr/e9aBuOcxNNzNzFQp2vaRMEju6q7xZMjYdcudUWL48vq9CoNa3X2ZArpqjkApR/TfYlG7glOj43NxuVDN4z9wIyUjaMHBfPgEhjaOaRyEFgEYITRwX1qDoXqcZdTVIq4Cn0uCH+Mvrz6Y+oUJGTJqH1k7N/DhzbSN9lJnVYaQW/yuvGHiGAwbb6Tcxiq2UqqhA9MfbPpmstDECs46v9Z3BT252KvYEQY7Q1v9g2gFhHvFGWISUxs80rnnPhEYfa11JoLvj2t8cowkE4pvj4OH32Eoyvc5H07hW3F5xpuF7Jt7N09TNZkUrpmiRJEhfrVNgjsrWO77/q5h8mXGd+9vYmz++yzKu+63x8x1MpeigGCG73Dpu9Otm5eydOZfpJ39ZfZWUb7G2JahgHaGweM9dmnpJtzHQgijmHjjfAx9jgnQ8IQz9nkFmyMI8H7HouwalnrJtpSSbvMqOQ0kiZhMzdBKH5pD3tjLgSlgA0pKelBwlooY6jGlj4LrtbDAxa6cZyXiFoqWpT1w=
```
Comment (dhermes) [inline .travis.yml]: Why does this keep changing? Why not just use the env. vars. in the UI?

File: system_tests/nox.py [hunk 14]
```diff
@@ -24,102 +24,124 @@
...
 
 
-def setup_cloud_sdk():
+def install_cloud_sdk(session):
     """Downloads and installs the Google Cloud SDK."""
+    # Configure environment variables needed by the SDK.
+    # This sets the config root to the tests' config root. This prevents
+    # our tests from clobbering a developer's configuration when running
+    # these tests locally.
+    session.env[CLOUD_SDK_CONFIG_ENV] = str(CLOUD_SDK_ROOT)
```
Comment (dhermes) [inline system_tests/nox.py]: This is putting the horse back in front of the cart?

File: system_tests/nox.py [hunk 15]
```diff
@@ -24,102 +24,124 @@
...
-    kwargs.setdefault('silent', True)
-    env = os.environ.copy()
-    env.update(kwargs.pop('env', {}))
-    Command(args, env=env, **kwargs).run()
+# The full path to the gcloud cli executable.
+GCLOUD = str(CLOUD_SDK_INSTALL_DIR.join('bin', 'gcloud'))
 
+# gcloud requires Python 2.7 and doesn't work on 3.x, so we need to tell it
+# where to find 2.7 when we're running in a 3.x environment.
+CLOUD_SDK_PYTHON_ENV = 'GCLOUD_PYTHON'
```
Comment (dhermes) [inline system_tests/nox.py]: Hey they stole our name!

File: system_tests/nox.py [hunk 16]
```diff
@@ -24,102 +24,124 @@
...
+    # these tests locally.
+    session.env[CLOUD_SDK_CONFIG_ENV] = str(CLOUD_SDK_ROOT)
+    # This tells gcloud which Python interpreter to use (always use 2.7)
+    session.env[CLOUD_SDK_PYTHON_ENV] = CLOUD_SDK_PYTHON
 
     # If the sdk already exists, we don't need to do anything else.
-    if CLOUD_SDK_ROOT.join('google-cloud-sdk').exists():
+    # Note that because of this we do not attempt to update the sdk -
+    # if the CLOUD_SDK_ROOT is cached, it will need to be periodically cleared.
+    if CLOUD_SDK_INSTALL_DIR.exists():
```
Comment (dhermes) [inline system_tests/nox.py]: `exists()` rather than `isdir()`?

File: system_tests/nox.py [hunk 17]
```diff
@@ -24,102 +24,124 @@
...
-    tar_path.remove()
+    session.run(
+        'tar', 'xzf', str(tar_path), '-C', str(CLOUD_SDK_ROOT))
+    session.run(tar_path.remove)
 
     # Run the install script.
-    prerun(
-        str(CLOUD_SDK_ROOT.join('google-cloud-sdk', 'install.sh')),
+    session.run(
+        str(CLOUD_SDK_INSTALL_DIR.join('install.sh')),
```
Comment (dhermes) [inline system_tests/nox.py]: For realz, why is this necessary?

File: system_tests/nox.py [hunk 18]
```diff
@@ -24,102 +24,124 @@
...
     # print-access-token
-    dest = CLOUD_SDK_ROOT.join('application_default_credentials.json')
-    if dest.exists():
-        dest.remove()
-    py.path.local(application_default_credentials).copy(dest)
+    def copy_credentials():
+        dest = CLOUD_SDK_ROOT.join('application_default_credentials.json')
+        if dest.exists():
+            dest.remove()
+        py.path.local(application_default_credentials).copy(dest)
```
Comment (dhermes) [inline system_tests/nox.py]: :-1: on the closure. Does `session.run` not allow you to pass args to a function? If not, why not define `copy_credentials(app_creds)` as a global and then use `functools.partial(copy_credentials, app_creds)` to make your callable with no args.

File: .travis.yml (hunk 13 above)
Comment (jonparrott) [inline .travis.yml]: I switched this to put the CLOUD_SDK_ROOT into the global env vars. It needs to stay here so the the `encrypt_secrets.sh` file can update this variable. (I also abhor UI as a the source of truth).

File: system_tests/nox.py (hunk 16 above)
Comment (jonparrott) [inline system_tests/nox.py]: Yes, that's fine, no need to be more specific.

File: system_tests/nox.py (hunk 14 above)
Comment (dhermes) [inline system_tests/nox.py]: Ha NVM. Was trying to make a (bad) joke

**Comment by jonparrott** [comment]
No idea why travis is failing on py35-system, as there doesn't seem to be any output from the failing commands. Investigating now.

**Comment by jonparrott** [comment]
V I C T O R Y
